### PR TITLE
Split Context and ProblemDescription (stage 3)

### PR DESCRIPTION
**Pull request by averinevg**
This PR splits `Context` and `ProblemDescription`.

### CODE REVIEW

**Comment by atamazov** [comment]
~900 lines of code removed 👍

**Comment by atamazov** [comment]
This needs extensive testing, including tuning modes

File: src/include/miopen/convolution.hpp [hunk 1]
```diff
@@ -169,17 +169,24 @@ struct ConvolutionDescriptor : miopenConvolutionDescriptor
...
+    ForwardBackwardGetWorkSpaceSizeImplicitGemm(const miopen::ConvolutionContext& ctx,
+                                                const miopen::ProblemDescription& problem) const;
 
     std::size_t
-    ForwardBackwardDataGetWorkSpaceSizeWinograd(const miopen::ConvolutionContext& ctx) const;
+    ForwardBackwardDataGetWorkSpaceSizeWinograd(const miopen::ConvolutionContext& ctx,
+                                                const miopen::ProblemDescription& problem) const;
 
-    bool IsWinograd3x3SupportedAndFast(miopen::ConvolutionContext& ctx) const;
+    bool IsWinograd3x3SupportedAndFast(const miopen::ConvolutionContext& ctx) const;
```
Comment (DrizztDoUrden) [inline src/include/miopen/convolution.hpp:187]: Should this have both signatures? Shouldn't the first go under `MIOPEN_CONV_CONTEXT_USE_FIN_COMPAT_API`?

**Review by DrizztDoUrden — APPROVED** [review]
LGTM, just one minor suggestion

File: src/include/miopen/conv/context.hpp [hunk 2]
```diff
@@ -48,28 +50,23 @@ struct ConvolutionContext : ExecutionContext
...
-    {
-    }
+#endif
+    explicit ConvolutionContext(const ExecutionContext& ctx) : ExecutionContext(ctx) {}
 
-    void SetupFloats();
+#if MIOPEN_CONV_CONTEXT_USE_FIN_COMPAT_API
+    void SetupFloats() { SetupFloats(problem); }
+#endif
+    void SetupFloats(const ProblemDescription& problem);
```
Comment (DrizztDoUrden) [inline src/include/miopen/conv/context.hpp:62]: I'd suggest moving it to `conv::ProblemDescription` as
```C++
void ProblemDescription::SetupFloats(ExecutionContext& ctx) const
```
To ease removal of ConvolutionContext later on.

File: src/include/miopen/convolution.hpp (hunk 1 above)
Comment (averinevg) [reply src/include/miopen/convolution.hpp:187]: @DrizztDoUrden You're right, the first one should be under `MIOPEN_CONV_CONTEXT_USE_FIN_COMPAT_API`. `context.hpp` is not included here and I decided not to include it as this code will be removed after making changes in the `Fin`. This macro present in the definition of the function.

File: src/include/miopen/conv/context.hpp (hunk 2 above)
Comment (averinevg) [reply src/include/miopen/conv/context.hpp:62]: @DrizztDoUrden I would suggest not making functional changes in this PR so as not to overload it.

**Comment by atamazov** [comment]
Performance/correctness testing is in progress...

File: src/include/miopen/conv/context.hpp (hunk 2 above)
Comment (DrizztDoUrden) [reply src/include/miopen/conv/context.hpp:62]: Okay, not a problem.

**Comment by atamazov** [comment]
## 🌀 Local performance & correctness testing

- HW/SW configurations
  - Radeon VII (gfx906/60), ROCm 5.3.0 + rocMLIR 1.1.0 + CK
  - Navi21, rocm/miopen:ci_mlir_tag_rocm-5.5.0
- Tested configs
  - 93 known FP32 convolutions
  - 93 known FP16 convolutions
- Library modes
  - Normal Find mode
  - Default Find mode (DYNAMIC_HYBRID)
  - Immediate Mode
- HIP backend
- branch a731bae81 vs develop 282d383b5

Results:
- 🟢 Navi 21 - OK
- 🟢 Radeon VII - IN PROGRESS

**Comment by atamazov** [comment]
Testing done, good to go!

**Comment by averinevg** [comment]
> Testing done, good to go!

@atamazov Thanks!

**Review by junliume — APPROVED** [review]

**Pull request merged**